apollo-runtime/src/main/java/com/apollographql/android/cache/ResponseReaderShadow.java

```diff
@@ -0,0 +1,30 @@
+package com.apollographql.android.cache;
+
+import com.apollographql.android.api.graphql.Field;
+import com.apollographql.android.api.graphql.Operation;
+
+import java.util.List;
+import java.util.Map;
+
+public interface ResponseReaderShadow {
+
+  void willResolveRootQuery(Operation operation);
+
+  void willResolve(Field field, Operation.Variables variables);
+
+  void didResolve(Field field, Operation.Variables variables);
+
+  void didParseScalar(Object value);
+
+  void willParseObject(Map<String, Object> objectMap);
+
+  void didParseObject(Map<String, Object> objectMap);
+
+  void didParseList(List array);
```

Comment:
iOS only uses `willParseList` to allocate more space on their "stack" data structure. I'm not aware of any similar pattern in Java (besides setting the size when a collection is first initialized)